packages/smithy-aws-core/src/smithy_aws_core/credentials_resolvers/credentials_resolver_chain.py

```diff
@@ -0,0 +1,54 @@
+from typing import Callable, List
+
+from smithy_aws_core.credentials_resolvers import EnvironmentCredentialsResolver
+from smithy_aws_core.identity import AWSCredentialsIdentity, AWSCredentialsResolver
+from smithy_core.aio.interfaces.identity import IdentityResolver
+from smithy_core.exceptions import SmithyIdentityException
+from smithy_core.interfaces.identity import IdentityProperties
+
+import os
+
+
+def _env_creds_available() -> bool:
+    return bool(os.getenv("AWS_ACCESS_KEY_ID")) and bool(
+        os.getenv("AWS_SECRET_ACCESS_KEY")
+    )
+
+
+def _build_env_creds() -> AWSCredentialsResolver:
+    return EnvironmentCredentialsResolver()
+
+
+type CredentialSource = tuple[Callable[[], bool], Callable[[], AWSCredentialsResolver]]
+_DEFAULT_SOURCES: list[CredentialSource] = [(_env_creds_available, _build_env_creds)]
+
+
+class CredentialsResolverChain(
+    IdentityResolver[AWSCredentialsIdentity, IdentityProperties]
+):
+    """Resolves AWS Credentials from system environment variables."""
+
+    def __init__(self, *, sources: List[CredentialSource] | None = None):
```

Comment:
Sources should be a `Sequence`. But also, why isn't it just a sequence of credential resolvers that take the same identity?